### PR TITLE
Declare listeners as fields in the component implementation.

### PR DESCRIPTION
**Pull request by spetrovic77**
This will allow us to build a static map of component -> listeners, which a deployer may read from a binary to initialize the listeners ahead of time (e.g., a Kube deployer).

Other changes:
  * Listeners are customized in the application config file.
  * The weavelet no longer knows about listener options.
  * Slight simplifying changes to the envelope protocol.

### CODE REVIEW

**Review by mwhittaker — APPROVED** [review]
Awesome change Srdjan!!! 👂👂👂

File: component.go [hunk 1]
```diff
@@ -175,29 +159,61 @@ func (r Ref[T]) Get() T { return r.value }
...
+//	  myOtherListener weaver.Listener
+//	}
+//
+// Listener options may be passed to the Listener in the application config.
+// For example, to specify local addresses for the above two listeners,
+// add the following lines to the application config file:
 //
-// A Listener implements the net.Listener interface, so you can use a Listener
-// wherever you use a net.Listener. For example,
+//	[listener.mylistener]
```
[the remark this answers is not in the document]
Comment (spetrovic77) [reply component.go:177]: We discussed this offline. This is definitely better, but only works for a single option.

File: examples/chat/server.go [hunk 2]
```diff
@@ -35,24 +34,19 @@ import (
...
 
 func (s *server) Main(ctx context.Context) error {
 	s.httpServer.Handler = instrument(s.label, s)
-	lis, err := s.Listener("chat", weaver.ListenerOptions{LocalAddress: *flagAddress})
-	if err != nil {
-		return err
-	}
-	s.Logger().Debug("Chat service available", "address", lis)
-	return s.httpServer.Serve(lis)
+	s.Logger().Debug("Chat service available", "address", s.chat.Addr())
```
[the remark this answers is not in the document]
Comment (spetrovic77) [reply examples/chat/server.go:48]: Good point. Fixed all of these instances.

File: filllisteners.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+		}
+		if ref.Type().Field(0).Name != "Listener" {
+			continue
+		}
+		if ref.Type().Field(1).Name != "proxyAddr" {
+			continue
+		}
+
+		// Listener name is a field name, unless a tag is present.
+		lisName := strings.ToLower(s.Type().Field(i).Name)
```
[the remark this answers is not in the document]
Comment (spetrovic77) [reply filllisteners.go:62]: I suppose that would be okay. That would name the listener "Listener", which isn't too bad. Can you put tags on embedded fields?

**Review by spetrovic77 — COMMENTED** [review]
PTAL!

File: filllisteners.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mwhittaker) [reply filllisteners.go:62]: Yup, you can put tags on embedded fields! Allowing the embeddings sounds good to me. I'd also be down with having `weaver generate` forbid embedded `weaver.Listener`s. I'm not sure which one is clearer.

File: filllisteners.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+		if tag := s.Type().Field(i).Tag.Get("weaver"); tag != "" {
+			lisName = strings.ToLower(tag)
```
[the remark this answers is not in the document]
Comment (mwhittaker) [reply filllisteners.go:64]: I think I have a slight preference for being case sensitive actually and letting `foo` and `Foo` be distinct. And maybe we can remove the restriction that listener names be DNS names actually? That almost seems like an implementation detail leaking into the API? I don't feel strongly though.

[the remark this answers is not in the document]
Comment (spetrovic77) [reply filllisteners.go:64]: Okay, let's discuss in the team meeting when Sanjay gets back. DNS names aside, I don't like the distinction between Foo and foo. I'll leave it as-is for now.

File: filllisteners.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (spetrovic77) [reply filllisteners.go:62]: Gotcha. I don't see a reason for us to forbid it.

File: filllisteners.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mwhittaker) [reply filllisteners.go:64]: Sounds good to me!

**Pull request merged**